mongodump/prepare.go

```diff
@@ -194,9 +196,26 @@ func (dump *MongoDump) outputPath(dbName, colName string) string {
 		root = dump.OutputOptions.Out
 	}
 
-	return filepath.Join(root, dbName, util.EscapeCollectionName(colName))
+	// Encode a new output path for collection names that would result in a file name greater
+	// than 255 bytes long. This includes the longest possible file extension: .metadata.json.gz
+	// The new format is <truncated-url-encoded-collection-name>%24<collection-name-hash-base64>
+	// where %24 represents a $ symbol delimiter (e.g. aVeryVery...VeryLongName%24oPpXMQ...).
+	if len(colName) > 238 {
```

Comment:
We need to escape the name before checking the length. Escaping can increase the length, so a collection name shorter than 238 bytes could result in a filename longer than 238 bytes.